src/circuit/gadget/sinsemilla/chip/hash_to_point.rs

```diff
@@ -0,0 +1,446 @@
+        let x_q = *Q.coordinates().unwrap().x();
+        let y_q = *Q.coordinates().unwrap().y();
+
+        // Initialize the accumulator to `Q`.
+        let (mut x_a, mut y_a): (X<pallas::Base>, Y<pallas::Base>) = {
+            // Constrain the initial x_q to equal the x-coordinate of the domain's `Q`.
+            let fixed_x_q =
+                region.assign_fixed(|| "fixed x_q", config.constants, offset, || Ok(x_q))?;
+            let x_q_cell = region.assign_advice(|| "x_q", config.x_a, offset, || Ok(x_q))?;
+            region.constrain_equal(&config.perm, fixed_x_q, x_q_cell)?;
+
+            // This cell gets copied into itself by the first call to `hash_piece` below.
+            let x_a = CellValue::new(x_q_cell, Some(x_q));
+
+            // Constrain the initial x_a, lambda_1, lambda_2, x_p using the fixed y_q
+            // initializer.
+            region.assign_fixed(|| "fixed y_q", config.fixed_y_q, offset, || Ok(y_q))?;
+
+            let y_a = Some(y_q);
+
+            (x_a.into(), y_a.into())
+        };
+
+        let mut zs_sum: Vec<Vec<CellValue<pallas::Base>>> = Vec::new();
+
+        // Hash each piece in the message except the final piece.
+        for piece in message[0..(message.len() - 1)].iter() {
+            // The value of the accumulator after this piece is processed.
+            let (x, y, zs) = self.hash_piece(region, offset, piece, x_a, y_a)?;
+
+            // Since each message word takes one row to process, we increase
+            // the offset by `piece.num_words` on each iteration.
+            offset += piece.num_words();
+
+            // Update the accumulator to the latest value.
+            x_a = x;
+            y_a = y;
+            zs_sum.push(zs);
+        }
+
+        // Hash the final message piece.
+        let y_a = {
+            let piece = &message[message.len() - 1];
+            // The value of the accumulator after this piece is processed.
+            let (x, y, zs) = self.hash_piece(region, offset, piece, x_a, y_a)?;
+
+            // Since each message word takes one row to process, we increase
+            // the offset by `piece.num_words` on each iteration.
+            offset += piece.num_words();
+
+            // Update the accumulator to the latest value.
+            x_a = x;
+            y_a = y;
+            zs_sum.push(zs);
+
+            // Assign and constrain the final `y_a`.
+            region.assign_fixed(
+                || "qs_2 = 2 on final row",
+                config.q_sinsemilla2,
+                offset - 1,
+                || Ok(pallas::Base::from_u64(2)),
+            )?;
+
+            let y_a_cell = region.assign_advice(
+                || "y_a",
+                config.lambda_1,
+                offset,
+                || y_a.ok_or(Error::SynthesisError),
+            )?;
+
+            // Assign lambda_2 and x_p zero values since they are queried
+            // in the gate.
```

Comment:
I don't understand this.

Reply:
The last gate queries these values, even though in this case they are unconstrained (since we are instead constraining the y<sub>A,n</sub> output). So we need to assign them to something, or else `MockProver` complains that an active gate isn't fully assigned.

Reply:
Oh, I see, they're multiplied by zero.

Reply:
```suggestion
            // Assign lambda_2 and x_p zero values since they are queried
            // in the gate. (The actual values do not matter since they are
            // multiplied by zero.)
```